### PR TITLE
[mlir][python] Add `walk` method to PyOperationBase

### PR DESCRIPTION
**Pull request by uenoku**
This commit adds `walk` method to PyOperationBase that uses a python object as a callback, e.g. `op.walk(callback)`. Currently callback must return a walk result explicitly. 

We(SiFive) have implemented walk method with python in our internal python tool for a while. However the overhead of python is expensive and it didn't scale well for large MLIR files. Just replacing walk with this version reduced the entire execution time of the tool by 30~40% and there are a few configs that the tool takes several hours to finish so this commit significantly improves tool performance.

### CODE REVIEW

**Comment by ftynse** [comment]
I wonder if this is something that should be implemented in python instead. We have access to op nesting structure of regions/blocks, so it shouldn't be hard. And not jumping between Python and C++ callbacks will simplify debugging for users.

**Review by uenoku — COMMENTED** [review]
Thank you for the feedback!

> We have access to op nesting structure of regions/blocks, so it shouldn't be hard. 

Yes exactly. We (SiFive) have implemented walk method in that way in our internal python tool for a while. However the overhead of python is expensive and it didn't scale well for large MLIR files witin SIFive. Just replacing `walk` with this version reduced the entire execution time of the tool by 30~40% and there are a few configs that the tool takes several hours to finish so we want to use the native walk method (though we should fix our tool algorithmically). 

I agree that debugging will be harder so I'm fine if we cannot upstream this. In that case we'll probably implement it in CIRCT as a helper function.

File: mlir/lib/Bindings/Python/IRCore.cpp [hunk 1]
```diff
@@ -1249,6 +1249,19 @@ void PyOperationBase::writeBytecode(const py::object &fileObject,
                               .str());
 }
 
+void PyOperationBase::walk(py::object callback, bool usePreOrder) {
```
Comment (ftynse) [inline mlir/lib/Bindings/Python/IRCore.cpp:1252]: I think it is possible to spell out the callback type as `std::function<void (MlirOperation)>` here, not sure if this will have any performance implications.

Consider exposing `MlirWalkResult` in C and Python API to allow functions to stop the walk early. 

I'm also not a fan of using a boolean kwarg for order, but can live with it.

File: mlir/test/python/ir/operation.py [hunk 2]
```diff
@@ -1015,3 +1015,36 @@ def testOperationParse():
...
+    print("Post-order")
+    module.operation.walk(callback)
+
+    # Test pre-order walk.
+    # CHECK-NEXT:  Pre-order
+    # CHECK-NEXT:  builtin.module
+    # CHECK-NEXT:  func.fun
+    # CHECK-NEXT:  func.return
+    print("Pre-order")
+    module.operation.walk(callback, True)
```
Comment (ftynse) [inline mlir/test/python/ir/operation.py:1050]: Could we test what happens if the callback raises an error?

File: mlir/lib/Bindings/Python/IRCore.cpp [hunk 3]
```diff
@@ -1249,6 +1249,19 @@ void PyOperationBase::writeBytecode(const py::object &fileObject,
...
+  MlirOperationWalkCallback walkCallback = [](MlirOperation op,
+                                              void *userData) {
+    py::object *fn = static_cast<py::object *>(userData);
+    (*fn)(op);
+  };
+  mlirOperationWalk(operation, walkCallback, &callback,
+                    usePreOrder ? MlirWalkOrder::MlirWalkPreOrder
+                                : MlirWalkOrder::MlirWalkPostOrder);
+}
+
```
Comment (joker-eph) [inline mlir/lib/Bindings/Python/IRCore.cpp:1264]: I would like the c++ walk to evolve to support two callbacks to have both pre/post order visitation in the same walk.

This isn't a blocker for your change, but I'll need to break the C API, so just a heads up :)
(I suspect we can preserve the Python API backward compatible on top of it)

**Comment by makslevental** [comment]
Just commenting to say thanks for adding this - I've been meaning to add this extension for a while and it got away from me.

File: mlir/lib/Bindings/Python/IRCore.cpp (hunk 1 above)
Comment (uenoku) [reply mlir/lib/Bindings/Python/IRCore.cpp]: Thank you for suggestion!  I changed callback type to `std::function<MlirWalkResult (MlirOperation)>` as well as adding `MlirWalkResult` CAPI and python enums for `MlirWalkResult/MlirWalkOrder`.

File: mlir/test/python/ir/operation.py (hunk 2 above)
Comment (uenoku) [reply mlir/test/python/ir/operation.py]: Added a test for exception. It seems working as expected.

File: mlir/include/mlir/Bindings/Python/PybindAdaptors.h [hunk 4]
```diff
@@ -18,6 +18,7 @@
 #ifndef MLIR_BINDINGS_PYTHON_PYBINDADAPTORS_H
 #define MLIR_BINDINGS_PYTHON_PYBINDADAPTORS_H
 
+#include <pybind11/functional.h>
```
Comment (uenoku) [inline mlir/include/mlir/Bindings/Python/PybindAdaptors.h:21]: This was needed to implicitly convert python object to std::function.

File: mlir/lib/Bindings/Python/IRCore.cpp [hunk 5]
```diff
@@ -1249,6 +1250,21 @@ void PyOperationBase::writeBytecode(const py::object &fileObject,
...
+void PyOperationBase::walk(
+    std::function<MlirWalkResult(MlirOperation)> callback,
+    MlirWalkOrder walkOrder) {
+  PyOperation &operation = getOperation();
+  operation.checkValid();
+  MlirOperationWalkCallback walkCallback = [](MlirOperation op,
+                                              void *userData) {
+    auto *fn =
+        static_cast<std::function<MlirWalkResult(MlirOperation)> *>(userData);
+    return (*fn)(op);
```
Comment (makslevental) [inline mlir/lib/Bindings/Python/IRCore.cpp:1262]: thought: can you actually check the result here to see if it's `py::none()` (i.e., the callback doesn't have a `return` statement) and then default to `MlirWalkResult::MlirWalkResultAdvance`. I think that's reasonable semantics? @ftynse

Comment (uenoku) [reply mlir/lib/Bindings/Python/IRCore.cpp:1262]: I'm not sure it works since the function is type casted to  `std::function<MlirWalkResult(MlirOperation)> callback`. 
To allow void functions I think we could change the callback type to `std::variant<std::function<MlirWalkResult(MlirOperation)>, std::function<void(MlirOperation)>` or maybe `py::function` (and dynamically type casts the returned type).

Comment (makslevental) [reply mlir/lib/Bindings/Python/IRCore.cpp:1262]: variant is overkill for sure but `py::function` might work? anyway it's just a thought rather than a hard request so nbd if you don't want to bother.

Comment (mikeurbach) [reply mlir/lib/Bindings/Python/IRCore.cpp:1262]: My two cents is if we can do this in a way that doesn't add much overhead it would be nice. But I'm also happy to see this go in with the user callback required to explicitly return a WalkResult... "explicit is better than implicit".

File: mlir/lib/Bindings/Python/IRCore.cpp [hunk 6]
```diff
@@ -2989,8 +3014,7 @@ void mlir::python::populateIRCore(py::module &m) {
            py::arg("binary") = false, kOperationPrintStateDocstring)
       .def("print",
            py::overload_cast<std::optional<int64_t>, bool, bool, bool, bool,
-                             bool, py::object, bool>(
-               &PyOperationBase::print),
+                             bool, py::object, bool>(&PyOperationBase::print),
```
Comment (mikeurbach) [inline mlir/lib/Bindings/Python/IRCore.cpp:3017]: nit: seems like unrelated format difference snuck in.

**Review by mikeurbach — APPROVED** [review]
I don't have any major concern with the implementation, this looks good to me.

File: mlir/lib/Bindings/Python/IRCore.cpp (hunk 6 above)
Comment (makslevental) [reply mlir/lib/Bindings/Python/IRCore.cpp:3017]: i don't think there's anything that can be done about this? presumably @uenoku just did a `clang-format` and this happened? ie if this is reverted then `clang-format` will complain here as well.

Comment (uenoku) [reply mlir/lib/Bindings/Python/IRCore.cpp:3017]: Yes, `Check code formatting` fails otherwise so I'll just keep it as is.

Comment (mikeurbach) [reply mlir/lib/Bindings/Python/IRCore.cpp:3017]: Huh weird, I guess that snuck in via a previous change.

File: mlir/lib/Bindings/Python/IRCore.cpp (hunk 5 above)
Comment (ftynse) [reply mlir/lib/Bindings/Python/IRCore.cpp:1262]: I guess I'm too traumatized by the C++ version of omitting the return from a function and getting a random segfault, so I'm on the side of always having an explicit return. But not objecting strongly.

File: mlir/lib/CAPI/IR/IR.cpp [hunk 7]
```diff
@@ -717,17 +717,34 @@ void mlirOperationMoveBefore(MlirOperation op, MlirOperation other) {
   return unwrap(op)->moveBefore(unwrap(other));
 }
 
+static mlir::WalkResult translateWalkResult(MlirWalkResult result) {
```
Comment (ftynse) [inline mlir/lib/CAPI/IR/IR.cpp:720]:
```suggestion
static mlir::WalkResult unwrap(MlirWalkResult result) {
```

this is the convention we use for all other types.

File: mlir/lib/CAPI/IR/IR.cpp [hunk 8]
```diff
@@ -717,17 +717,34 @@ void mlirOperationMoveBefore(MlirOperation op, MlirOperation other) {
...
+
 void mlirOperationWalk(MlirOperation op, MlirOperationWalkCallback callback,
                        void *userData, MlirWalkOrder walkOrder) {
   switch (walkOrder) {
 
   case MlirWalkPreOrder:
     unwrap(op)->walk<mlir::WalkOrder::PreOrder>(
-        [callback, userData](Operation *op) { callback(wrap(op), userData); });
+        [callback, userData](Operation *op) {
+          return translateWalkResult(callback(wrap(op), userData));
```
Comment (ftynse) [inline mlir/lib/CAPI/IR/IR.cpp:740]:
```suggestion
          return unwrap(callback(wrap(op), userData));
```

File: mlir/lib/CAPI/IR/IR.cpp [hunk 9]
```diff
@@ -717,17 +717,34 @@ void mlirOperationMoveBefore(MlirOperation op, MlirOperation other) {
...
+        });
     break;
   case MlirWalkPostOrder:
     unwrap(op)->walk<mlir::WalkOrder::PostOrder>(
-        [callback, userData](Operation *op) { callback(wrap(op), userData); });
+        [callback, userData](Operation *op) {
+          return translateWalkResult(callback(wrap(op), userData));
```
Comment (ftynse) [inline mlir/lib/CAPI/IR/IR.cpp:746]:
```suggestion
          return unwrap(callback(wrap(op), userData));
```

**Review by ftynse — APPROVED** [review]

File: mlir/lib/Bindings/Python/IRCore.cpp (hunk 5 above)
Comment (makslevental) [reply mlir/lib/Bindings/Python/IRCore.cpp:1262]: > I guess I'm too traumatized by the C++ version of omitting the return from a function and getting a random segfault

I am similarly traumatized but in Python-land I think it's fairly well-known that a function with no return statement actually returns `None`.

**Comment by uenoku** [comment]
Thank you for all comments! I think the current implementation that requires explicit return results from python callback would be good start so I'd like to just merge the PR once CI passed. Let me open a tracking issue for allowing void function in the callback.

**Comment by metaflow** [comment]
Hi, we see a `libc++abi: terminating due to uncaught exception of type pybind11::error_already_set: ValueError: <EMPTY MESSAGE>` in the mlir/test/python/ir/operation.py

**Comment by uenoku** [comment]
> 
Hi, we see a libc++abi: terminating due to uncaught exception of type pybind11::error_already_set: ValueError: <EMPTY MESSAGE> in the mlir/test/python/ir/operation.py

@metaflow Sorry I missed the message. It looks that the issue was fixed by https://github.com/llvm/llvm-project/pull/89225. Thank you @tomnatan30!
